c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -208,18 +208,31 @@ impl<'a, 'tcx> UnlowerVisitor<'a, 'tcx> {
 
         // Most exprs end with an assignment, storing the result into a temporary.
         match ex.kind {
-            hir::ExprKind::Assign(pl, rv, _span) => {
+            hir::ExprKind::Assign(pl, _rv, _span) => {
                 // For `Assign`, we expect the assignment to be the whole thing.
-                let (loc, _mir_pl, _mir_rv) = match self.get_sole_assign(&locs) {
+                let (loc, _mir_pl, mir_rv) = match self.get_sole_assign(&locs) {
                     Some(x) => x,
                     None => {
                         warn("expected exactly one StatementKind::Assign");
                         return;
                     }
                 };
+                let desc = match mir_rv {
+                    mir::Rvalue::Use(op) => op
+                        .place()
+                        .map(|pl| {
+                            if is_var(pl) && self.mir.local_kind(pl.local) == mir::LocalKind::Temp {
+                                MirOriginDesc::LoadFromTemp
+                            } else {
+                                MirOriginDesc::Expr
+                            }
+                        })
+                        .unwrap_or(MirOriginDesc::Expr),
+                    _ => MirOriginDesc::Expr,
+                };
```

Comment:
This duplicates the `op_is_temp` logic from `record_operand` - would it make sense to factor that out?

Reply:
Added in 8172666f

Reply:
I was thinking more like `fn op_is_temp(&self, op: &Operand<'tcx>) -> bool`, or maybe `fn operand_desc(&self, op: &Operand<'tcx>) -> MirOriginDesc`.  I'm not a fan of the `get_desc` from 8172666 because I don't think its logic makes sense outside of the specific context of `op_is_temp`.  For example, when looking at a `StatementKind::Assign` that stores into a temporary, calling `get_desc` on the LHS `Place` would produce `LoadFromTemp`, which seems wrong.

Reply:
I was thinking more like `fn op_is_temp(&self, op: &Operand<'tcx>) -> bool`, or maybe `fn operand_desc(&self, op: &Operand<'tcx>) -> MirOriginDesc`.  The current `get_desc` doesn't entirely make sense outside the specific context of `op_is_temp`.  For example, if you have a `StatementKind::Assign` that stores into a temporary, calling `get_desc` on the LHS would return `LoadFromTemp`, which probably isn't what you'd want.

Reply:
5cc3fb7e